### DISCUSSION TITLE
$14 Setting Safety Door bit invert doesn't stick

### DISCUSSION THREAD

**Discussion by jerseyguy1996** [grblHAL/core#23] [General]
On ioSender 2.0.33 Settings: Grbl tab as well as in the console tab of Lightburn, setting $14=78 should result in Feed Hold, Cycle Start, Safety Door, and Estop signals inverted.  After saving and then reloading, the setting reverts to $14=70 with the Safety Door signal invert reverting to default (non-inverted) value.  Tried setting Feed Hold, Cycle Start, Estop, and Reset to invert ($14-71) and after reloading that stayed correctly at $14=71.  Adding Safety Door inverted so $14=79 and after reloading it reverted back to $14 = 71 so it seems to be specific to the Safety Door bit.

Edit:  This is using the current master branch of the grblHAL iMXRT1062 repo with SD Card and Ethernet enabled

**Comment by terjeio** [comment]
Have you enabled the safety door input handling? It is disabled by default.

https://github.com/grblHAL/core/blob/edd243ac8a504e0f658630fc92a91ca3cf9375b1/config.h#L73-L77

If not the setting will not stick as inverting the input might otherwise cause problems.

**Reply by jerseyguy1996** [reply]
I did not enable the safety door input.  I am not at the shop now but I have no doubt that that is the issue.  Thank you so much for the response and for creating this fantastic software!

**Comment by phil-barrett** [comment]
Coming a bit late to this.  Was playing with enabling Door and ran into this.  grblHAL after Feb 2021 no longer allows setting bits in $14 that don't an associated capability.  (I may be describing this incorrectly but that is the effect I am seeing).  I wanted to set Door to NO from it's default NC so I can play with it on a bare Teensy. But doing that with a build with door disabled didn't work.  While I get that ioSender might want to disallow setting the bits so as not to confuse uses about a non-existent capability, even using a terminal app, the bits don't get set.  And of course, grblHAL refuses to allow $nn=mm with the door "open" so there is no way to change it to NO via a sender/terminal app.

To be clear in most cases you just short the door input to ground and everyone is happy.  But on my Pro board I need 12V to make the I/O work and that wasn't convenient at the time. I think this is a problem for someone setting up a grblHAL system - bench testing is common and that may not have 12V initially.

So, my question is why does grblHAL need to prevent unused bits from being set?

**Reply by terjeio** [reply]
> So, my question is why does grblHAL need to prevent unused bits from being set?

It is to avoid confusion over which functionality is available. The current ioSender version labels them N/A in the settings tab and the [latest edge version](http://www.io-engineering.com/downloads) hides them.

> But doing that with a build with door disabled didn't work.

The door input pin is not read if disabled, it should not show up in the status report. It is on my todo list to add it to aux inputs when disabled.

>  And of course, grblHAL refuses to allow $nn=mm with the door "open" so there is no way to change it to NO via a sender/terminal app.

It should not be neccesary to change the setting when it is hidden, if there is a `D` showning up in the `Pn:` element in the status report then that is a bug. Does it?

**Reply by terjeio** [reply]
>  But on my Pro board I need 12V to make the I/O work and that wasn't convenient at the time.

I just feed 5V to the 12V input from the Relay voltage selector, works for me. I do the same for the Pico board.

**Reply by phil-barrett** [reply]
There are no bugs here (which is why it's in discussions).  Yes, I understand your points.  My point was bench testing a Pro board without 12V requires that you have Door disabled in your build as there is no way to flip the door bit in $14 with it enabled or disabled. Pretty sure a few builders sure that will hit that. Not the end of the world but kind of clumsy and probably generates a support request.

**Reply by phil-barrett** [reply]
So you use a jumper from the 5V pin to the 12V header on the lower left. I am surprised the optos work with current that low on the Pro (about 1 mA). Does the LED light up?

**Reply by terjeio** [reply]
> My point was bench testing a Pro board without 12V requires that you have Door disabled in your build as there is no way to flip the door bit in $14 with it enabled or disabled.

Sorry - I do not get it, I am a bit tired today. You are able to flip it when Door input is enabled, when disabled there is no input to flip. You mean that will cause support requests? If so the same will be true for settings that does not change any behaviour when changed?
Or do you want to test the door input functionality even when it is disabled by using the setting to simulate the input?

**Reply by terjeio** [reply]
>  I am surprised the optos work with current that low on the Pro (about 1 mA). Does the LED light up?

It does, only a little less bright than the 5V LED. The input pin goes to 70mV.

**Reply by phil-barrett** [reply]
> Sorry - I do not get it, I am a bit tired today. You are able to flip it when Door input is enabled, when disabled there is no input to flip. You mean that will cause support requests? If so the same will be true for settings that does not change any behaviour when changed?
> Or do you want to test the door input functionality even when it is disabled by using the setting to simulate the input?

I was trying to be a bit too efficient.  With a build that has Door disabled, you can't invert the $14 bit to make the input NO.  With Door enabled, the pin is, by default, set up for NC - you can not get ioSender to run.  Via a terminal app, grblHAL won't let you issue a $14=xx and $X doesn't clear the condition.

**Reply by terjeio** [reply]
Ok, maybe I get it now - what is buggering you is that it is no longer possible to set $14 to invert the input when the door is disabled and then enable it. This to avoid the alarm on the following restart.

Perhaps it should always be possible to issue settings commands, even when in critical alarm modes?

**Reply by phil-barrett** [reply]
Yes, that would be helpful.  One sees it with Halt/Reset also.
